components/provisioner/internal/operations/stages/hibernation/wait_for_hibernation.go

```diff
@@ -0,0 +1,63 @@
+package hibernation
+
+import (
+	"context"
+	"time"
+
+	gardener_types "github.com/gardener/gardener/pkg/apis/core/v1beta1"
+	"github.com/kyma-project/control-plane/components/provisioner/internal/model"
+	"github.com/kyma-project/control-plane/components/provisioner/internal/operations"
+	"github.com/sirupsen/logrus"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+//go:generate mockery -name=GardenerClient
+type GardenerClient interface {
+	Get(ctx context.Context, name string, options v1.GetOptions) (*gardener_types.Shoot, error)
+}
+
+type WaitForHibernation struct {
+	gardenerClient GardenerClient
+	nextStep       model.OperationStage
+	timeLimit      time.Duration
+}
+
+func NewWaitForHibernation(gardenerClient GardenerClient, nextStep model.OperationStage, timeLimit time.Duration) *WaitForHibernation {
+	return &WaitForHibernation{
+		gardenerClient: gardenerClient,
+		nextStep:       nextStep,
+		timeLimit:      timeLimit,
+	}
+}
+
+func (c *WaitForHibernation) Name() model.OperationStage {
+	return model.WaitForHibernation
+}
+
+func (c *WaitForHibernation) TimeLimit() time.Duration {
+	return c.timeLimit
+}
+
+func (c *WaitForHibernation) Run(cluster model.Cluster, operation model.Operation, log logrus.FieldLogger) (operations.StageResult, error) {
+
+	log.Debugf("Starting WaitForHibernation stage for %s ...", cluster.ID)
+	shoot, err := c.gardenerClient.Get(context.Background(), cluster.ClusterConfig.Name, v1.GetOptions{})
+	if err != nil {
+		return operations.StageResult{}, err
+	}
+
+	if shoot.Status.IsHibernated {
+		log.Debugf("Cluster: %s is hibernated, proceeding to the next stage ...", cluster.ID)
+		return operations.StageResult{
+			Stage: c.nextStep,
+			Delay: 0,
+		}, nil
+	}
+
```

Comment:
Maybe it can be worthy to also add checking of `shoot.Status.LastOperation` to verify if the hibernation process  has not failed and in that case fail the step with `NonRecoverableError`. 

For example similar checking during cluster upgrade

```
	if shoot.Status.LastOperation.State == gardencorev1beta1.LastOperationStateFailed {
		err := fmt.Errorf(fmt.Sprintf("Gardener Shoot cluster upgrade failed. Last Shoot state: %s, Shoot description: %s", shoot.Status.LastOperation.State, shoot.Status.LastOperation.Description))
		return operations.StageResult{}, operations.NewNonRecoverableError(err)
	}
```